lib/sync/endpoint.php

```diff
@@ -0,0 +1,210 @@
+<?php
+/**
+ * Adds an http based signaling server for real time collaboration.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Reads the contents of $fd and returns them unserialized.
+ *
+ * @access private
+ * @internal
+ *
+ * @param integer $fd      A file descriptor.
+ *
+ * @return array Unserialized contents of fd.
+ */
+function _gutenberg_get_contents_from_file_descriptor( $fd ) {
+	$contents_raw = stream_get_contents( $fd );
+	$result       = array();
+	if ( $contents_raw ) {
+		$result = unserialize( $contents_raw );
+	}
+	return $result;
+}
+
+
+/**
+ * Makes the file descriptor content of $fd equal to the serialization of content.
+ * Overwrites what was previously in $fd.
+ *
+ * @access private
+ * @internal
+ *
+ * @param integer $fd      A file descriptor.
+ * @param array   $content Content to be serialized and written in a file descriptor.
+ */
+function _gutenberg_save_contents_to_file_descriptor( $fd, $content ) {
+	rewind( $fd );
+	$data = serialize( $content );
+	fwrite( $fd, $data );
+	ftruncate( $fd, strlen( $data ) );
+}
+
+/**
+ * Handles HTTP requests for the gutenberg_signaling_server action.
+ *
+ * @access private
+ * @internal
+ */
+function _gutenberg_wp_ajax_signaling_server() {
+	session_write_close();
+	$subscriber_to_messages_path = sys_get_temp_dir() . DIRECTORY_SEPARATOR . 'subscribers_to_messages.txt';
```

Comment:
We could use the more flexible `get_temp_dir()` instead of `sys_get_temp_dir()`

Reply:
Good one I did not knew this function existed 👍